src/aegis/src/aegis_server.erl

```diff
@@ -286,8 +278,10 @@ is_key_fresh(UUID) ->
     Now = fabric2_util:now(sec),
 
     case ets:lookup(?KEY_CHECK, UUID) of
-        [{UUID, ExpiresAt}] when ExpiresAt >= Now -> true;
-        _ -> false
+        [{UUID, ExpiresAt}] when ExpiresAt - ?KEY_EARLY_EXPIRE_SEC > Now ->
```

Comment:
You lost me here - so this makes keys to expire 5 secs earlier, i.e. this is virtually the same as reducing `[aegis] cache_max_age_sec` config parameter on 5 sec less. Expired keys and missing from cache keys are treated the same way by doing sync call, so what the point of this change?

Reply:
same question here. In epep we had a 'fresh', 'stale' and 'expired' category for this kind of thing (where a stale entry was usable until expired, and we forced a refresh when we first notice its stale). Did I really forget to do the same thing here? :(

Reply:
Aegis cache's refresh needs to read/write from fdb, hence it needs fdb context and can't be run on background off the  "parent" process, so 'stale' strategy doesn't really work here.

Reply:
> this is virtually the same as reducing [aegis] cache_max_age_sec config parameter on 5 sec less

I don't think so, because `ExpiresAt` is being compared to `now()` in both `is_key_fresh` and also `remove_expired_entries`, whereas this change basically says that the key is not "fresh" if it's 5 seconds away from expiring.  I think the `badmatch` crash exposed this race condition, and `KEY_EARLY_EXPIRE_SEC` is an attempt to minimize the impact of that race by not trying to use a cached key that's too close to expiring. I wonder if 10 seconds would be better though since that's the frequency used by the expiration reaper?

> Did I really forget to do the same thing here? :(

I think so, but that is effectively what adding this parameter does: `is_key_fresh` returns false not only when the key is expired, but also when it is stale. Maybe a better name would be `KEY_STALE_SEC`?

Reply:
So there is a race between `is_key_fresh` and `lookup` when `remove_expired_entries` called at the same second as `is_key_fresh`? Wouldn't then changing ` MatchConditions = [{'=<', '$1', fabric2_util:now(sec)}],`  in `remove_expired_entries` work better from logical point of view, because right now you are not really refreshing a key, but expiring it ahead of the configured period and a client will just experience a latency bump earlier, while really the intent here is to delay cache eviction.

This will also address a potential surprise in case "cache_max_age_sec"  happened to be set lower than `?KEY_EARLY_EXPIRE_SEC` and cache will stop working.

Reply:
Yeah, I think something like
```erlang
    MatchConditions = [{'=<', '$1', fabric2_util:now(sec) + ?KEY_EXPIRE_GRACE_SEC}],

```
would also work, and is likely better for the reasons you mention.

> you are not really refreshing a key, but expiring it ahead of the configured period and a client will just experience a latency bump earlier

By returning false from `is_key_fresh`, it effectively forces a refresh, but yes, it sneakily shortens `CACHE_MAX_AGE_SEC` by a few seconds. Seems like just adding a grace period where the cache is still honored is maybe more intuitive.

Reply:
As I see it "refresh" is different from "expire" is that it's happening transparently for the client, i.e. async and in background. Here both behave the same, because fdb context limitation, so from user's perspective it'll looks like they set expiration for 10 sec and got key expired every 5 sec instead, doesn't look intuitive.

Reply:
I pushed a fixup that reverts changes to `is_key_fresh` and instead honors a new configurable grace period before deleting entries in `remove_expired_entries`. It also required a tweak to one of the lru tests.